### PR TITLE
Add localization (again)

### PR DESCRIPTION
**Pull request by pardoman**
## Description
This PR picks up where #30 left off. The PR solves Issue #14.

The PR includes all changes from @HSZemi's branch: [spike/all-the-locales](https://github.com/SiegeEngineers/aoe2techtree/tree/spike/all-the-locales).

## New dependencies
- `i18next` - the core localization library
- `i18nextXHRBackend` - a plugin for i18next, which requests localization json files from the server.

Information on these 2 dependencies can be found in https://www.i18next.com/

## How to use
Simple UI has been implemented to switch between different languages. For now, only a few buttons have been added:
image

## Other UI Changes
- The civ dropdown lo longer displays a civ name until the data files get loaded.
- I took the liberty of adding a loading spinner that renders while the JSON files are being loaded, using the same color scheme as other UI elements. The code for the loading spinner comes from here: https://loading.io/css/
civ-spinner

## Known issues
The help tooltip is partially localized.  The code that  builds the tooltip grabs the main tooltip text from locales/<lang>/strings.json, but then further changes the text using a bunch of `string.replace()` calls that I've updated to localize. However, these locales are not present in `strings.json`, and thus I've added those strings manually into `common.json`. Only English and Spanish locales have these strings for now.

### CODE REVIEW

**Comment by pardoman** [comment]
I also did a separate PR which further updates the localization UI. Please check it out here:

https://github.com/pardoman/aoe2techtree/pull/1

File: index.html [hunk 1]
```diff
@@ -571,6 +630,122 @@
...
+     * Responsible for updating the site's locales (strings).
+     *
+     * @param {Array[]} err - populated if there's an error settings the language, else null
+     * @param {function} _t - A reference to function: i18next.t
+     */
+    function onLanguageSet(err, _t) {
+        
+        document.getElementById('loading').style.display = 'none';
+        
+        // Debut output
```
Comment (HSZemi) [inline index.html:719]: Clearly, that does not have to be emphasized by a comment

File: index.html [hunk 2]
```diff
@@ -571,6 +630,122 @@
...
+                    domElem.innerHTML = _t(locKey);
+                    break;
+
+                case "IMG":
+                    domElem.title = _t(locKey);
+                    
+            }
+        }
+
+        // Udpate SVG text nodes
```
Comment (HSZemi) [inline index.html:751]:
```suggestion
        // Update SVG text nodes
```

File: index.html [hunk 3]
```diff
@@ -60,11 +60,24 @@
...
-                    <!-- Consider updating the UI to display flags instead of text -->
-                    [<a href="#" title="English" data-lang="en">en</a>]
-                    [<a href="#" title="Spanish" data-lang="es">es</a>]
-                    [<a href="#" title="French"  data-lang="fr">fr</a>]
-                    [<a href="#" title="Chinese" data-lang="zh">zh</a>]
+                 <!-- "Language supported" in the same order as they appear in Microsoft's product page:
+                        https://www.microsoft.com/en-us/p/age-of-empires-ii-definitive-edition/9njdd0jgpp2q -->
+                    [<a href="#" title="English (United States)" data-lang="en">en</a>]
+                    [<a href="#" title="中文(中国)" data-lang="zh">zh-Hans</a>]
+                    [<a href="#" title="中文(香港特別行政區)" data-lang="tw">zh-Hant</a>]
```
Comment (HSZemi) [inline index.html:67]:
```suggestion
                    [<a href="#" title="中文(中国)" data-lang="zh">zh</a>]
                    [<a href="#" title="中文(香港特別行政區)" data-lang="tw">tw</a>]
```

File: index.html [hunk 4]
```diff
@@ -16,57 +16,65 @@
...
                 <p id="civtext"></p>
+                <div id="loading" class="lds-dual-ring"></div>
             </div>
             <div class="info">
+                <div class="info__margin">
+                    <!-- Consider updating the UI to display flags instead of text -->
+                    [<a href="#" title="English" data-lang="en">en</a>]
+                    [<a href="#" title="Spanish" data-lang="es">es</a>]
+                    [<a href="#" title="French"  data-lang="fr">fr</a>]
+                    [<a href="#" title="Chinese" data-lang="zh">zh</a>]
```
[the remark this answers is not in the document]
Comment (HSZemi) [reply index.html]: I see that that has already been taken care of

File: index.html [hunk 5]
```diff
@@ -406,16 +466,25 @@
...
+                    }
+                }
+                domSelect.selectedIndex = index;
+                loadCiv();
             });
 
             let xRefImage = document.createElement('img');
 
             xRefImage.src = `./img/Civs/${civ.toLowerCase()}.png`;
             xRefImage.title = `${civ}`;
```
Comment (HSZemi) [inline index.html:484]: This needs to be translated as well

**Comment by HSZemi** [comment]
The carets that get a special treatment still have some issues:
- The second town center is not translated at all
- After changing the language, the unique units and unique techs at the castle only get updated once the civilisation is changed

File: index.html [hunk 6]
```diff
@@ -406,16 +466,25 @@
         for (let civ of Object.keys(data.civ_names)) {
             let xRefLink = document.createElement('button');
             xRefLink.addEventListener('click', function() {
-              document.getElementById('civselect').value=civ;
-              loadCiv();
+                const domSelect = document.getElementById('civselect');
+                var index;
+                for (var i=0; i<domSelect.children.length; ++i) {
+                    if (domSelect.children[i].getAttribute('data-i18n')===civ) {
```
Comment (HSZemi) [inline index.html:472]: Bugfix:
```suggestion
                    if (domSelect.children[i].getAttribute('data-key')===civ) {
```

File: data/locales/en/common.json [hunk 7]
```diff
@@ -0,0 +1,26 @@
+{
+    "key": "Hello in English",
```
[the remark this answers is not in the document]
Comment (pardoman) [reply data/locales/en/common.json:2]: Good find, it was for testing.

File: index.html [hunk 8]
```diff
@@ -16,57 +16,65 @@
...
                 </div>
                 <div class="info__margin">
-                    Contribute on <a href="https://github.com/SiegeEngineers/aoe2techtree" target="_blank">GitHub</a>
+                    <span>Contribute on</span> <a href="https://github.com/SiegeEngineers/aoe2techtree" target="_blank">GitHub</a>
                 </div>
                 <div>
-                    Made by hszemi, Anda, exterkamp, paulirish,<br>
-                    with thanks to Jineapple, TriRem, pip, and NkoDragaš<br>
+                    <span>Made by</span> hszemi, Anda, exterkamp, paulirish,<br>
+                    <span>with thanks to</span> Jineapple, TriRem, pip, and NkoDragaš<br>
```
[the remark this answers is not in the document]
Comment (pardoman) [reply index.html:90]: Not any more. I initially had them localized, but decided to remove it. Will revert these changes.

File: index.html [hunk 9]
```diff
@@ -172,10 +192,19 @@
...
                         id: `${caret.id}_bg`
                     }).move(caret.x, caret.y);
-                    let name = formatName(caret.name);
+                    
+
+                    let entitytype = caret.getDataType();
+                    let langData = data.data[entitytype][caret.data_id];
+                    let name = caret.name;
+                    let languageNameId = "NONE";
+                    if (langData) {
```
[the remark this answers is not in the document]
Comment (pardoman) [reply index.html:213]: I don't think we should expect this to be null. However, I encountered this with one caret entry.  At the time, decided to patch and move on. I can leave a comment and look into it further next.

File: index.html [hunk 10]
```diff
@@ -219,15 +257,20 @@
...
-        civ(selectedCiv, tree);
-        if (selectedCiv in data.civ_helptexts) {
-            document.getElementById('civtext').innerHTML = data.strings[data.civ_helptexts[selectedCiv]];
-            document.getElementById('civlogo').src = `./img/Civs/${selectedCiv.toLowerCase()}.png`;
-            window.location.hash = selectedCiv;
+        const domSelect = document.getElementById('civselect');
+        const domCivText = document.getElementById('civtext');
+        const domCivLogo = document.getElementById('civlogo');
+
+        const dataKey = domSelect.children[domSelect.selectedIndex].getAttribute('data-key'); // i.e: "Aztects"
```
[the remark this answers is not in the document]
Comment (pardoman) [reply index.html:276]: I'll fix it on my end, thanks.

File: index.html [hunk 11]
```diff
@@ -219,15 +257,20 @@
...
+        civ(dataKey, tree);
+        if (dataKey in data.civ_helptexts) {
```
[the remark this answers is not in the document]
Comment (pardoman) [reply index.html:278]: IIRC it was needed during initial load

File: index.html (hunk 2 above)
Comment (pardoman) [reply index.html:751]: I'll fix it on my end.

File: index.html (hunk 5 above)
Comment (pardoman) [reply index.html:484]: Good catch.

File: index.html (hunk 3 above)
Comment (pardoman) [reply index.html:67]: Will fix it on my end.

**Comment by pardoman** [comment]
Thanks @HSZemi for the review! Updated the PR and will look into a few of the other lingering issues.

File: index.html (hunk 11 above)
[the remark this answers is not in the document]
Comment (pardoman) [reply index.html:279]: I'll double check.

[the remark this answers is not in the document]
Comment (pardoman) [reply index.html:279]: It ain't needed. Will remove.

**Comment by pardoman** [comment]
> * The second town center is not translated at all

Fixed with my latest commit.

> * After changing the language, the unique units and unique techs at the castle only get updated once the civilisation is changed

There are still 4 carets, one for each unique unit/tech, which I still need to get fixed. I left a TODO for the moment.

**Comment by HSZemi** [comment]
I updated the `spike/all-the-locales` branch, it was missing fixes for the help texts of Slinger and Imperial Camel Rider that I already did on master.

**Comment by pardoman** [comment]
Alright, I'll bring those changes in, too. Thanks.

**Comment by pardoman** [comment]
Rebasing is needed. I'll get back to this as soon as possible.

**Comment by HSZemi** [comment]
> Rebasing is needed. I'll get back to this as soon as possible.

Great, thank you!

**Comment by pardoman** [comment]
@HSZemi Hey, I created a new branch in my fork that contains the content of this PR plus latest master changes: https://github.com/pardoman/aoe2techtree/tree/pardoman/localization

I need help with a couple of things:
1. When merging, I had to resolve conflicts in `scripts/generateDataFiles.py`. Could you check to see if that is in okay shape?
2. I'd like to have the python script run to generate the updated data.json and strings.json files

Thanks

**Comment by pardoman** [comment]
Branch updated. I still need help with the 2 above items, @HSZemi . Thanks!

**Comment by HSZemi** [comment]
Sorry, I have been slacking. The merged script looks good. Here are the updated files: https://github.com/SiegeEngineers/aoe2techtree/commit/7fc9f6571f9fcc91de3a821722ea343156fe26cb

**Comment by pardoman** [comment]
Thanks for the updates! I got them merged successfully.

Given the amount of changes the branch has been through since I initially create the PR, I will close it and make a new one from a fresh branch.

**Comment by pardoman** [comment]
New PR: #48

**Pull request closed without merging**